### ISSUE TITLE
Returning int from an async function typed FutureOr<int> fails in dart2js

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#33639]
Dart -dev.64.1 and -dev.65

The following test succeeds with the VM and fails with dart2js with `TypeErrorImplementation`

```dart
import 'dart:async';
import 'package:test/test.dart';

void main() {
  test('sample', () async {
    var result = await _something();
    expect(result, 42);
  });
}

FutureOr<int> _something() async => 42;
```

```
  Async function returned a Future, was declared to return a FutureOr<int>.
  org-dartlang-sdk:///sdk/lib/_internal/js_runtime/lib/js_helper.dart 1764:37   Object.wrapException
  org-dartlang-sdk:///sdk/lib/_internal/js_runtime/lib/js_rti.dart 646:3        Object.throwTypeError
  test/sample_test.dart 16:37                                                   <fn>
  org-dartlang-sdk:///sdk/lib/_internal/js_runtime/lib/async_patch.dart 335:19  _wrapJsFunctionForAsync.closure.$protected
  package:stack_trace                                                           StackZoneSpecification._registerBinaryCallback.<fn>
  org-dartlang-sdk:///sdk/lib/_internal/js_runtime/lib/async_patch.dart 251:3   Object._asyncStartSync
  test/sample_test.dart 16:37                                                   Object._something
```

Source issue: https://github.com/dart-lang/shelf_test_handler/issues/5

**Comment by sigmundch** [comment]
@kevmoo I cannot repro this locally. I tried both 64.1 and 65. What am I doing wrong?

**Comment by sigmundch** [comment]
Fix underway in https://dart-review.googlesource.com/c/sdk/+/62442

The problem was related to the `--checked` flag. Dart2js prints a warning if you are in Dart2 and pass `--checked`, in that case it should only enable assertions (for now). When dart2js switched the default to Dart2, we started processing the `--checked` flag incorrectly and all the sudden, we turned on Dart1 checked-mode checks in the generated code. Dart1 doesn't understand FutureOr, so the extra check automatically added a type-error at compile-time.

The CL above fixes how we handle the flag, but long term the flag will be a noop, so `package:test` should be updated to pass `--enable-asserts` instead.
